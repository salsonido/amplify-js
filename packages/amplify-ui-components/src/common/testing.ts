import {
	boolean as booleanKnob,
	select as selectKnob,
	text as textKnob,
} from '@storybook/addon-knobs';

export const pixelThreshold = 10;

export const selectors = {
	// amplify-form-field
	formFieldDescription: '[data-test=form-field-description]',
};

export const knobs = {
	labelKnob(labelPlaceholder: string) {
		return textKnob('Label', labelPlaceholder);
	},

	descriptionKnob(descriptionPlaceholder: string) {
		return textKnob('Description', descriptionPlaceholder);
	},

	hintKnob(hintPlaceholder: string) {
		return textKnob('Hint', hintPlaceholder);
	},

	inputTypeKnob(inputTypePlaceholder: string) {
		return selectKnob(
			'Type',
			[
				'date',
				'email',
				'number',
				'password',
				'search',
				'tel',
				'text',
				'url',
				'time',
			],
			inputTypePlaceholder
		);
	},

	placeholderKnob(placeholderPlaceholder: string) {
		return textKnob('Placeholder', placeholderPlaceholder);
	},

	buttonTypeKnob(buttonTypePlaceholder: string) {
		return selectKnob(
			'Button type',
			['button', 'submit', 'reset'],
			buttonTypePlaceholder
		);
	},

	buttonTextKnob(buttonTextPlaceholder: string) {
		return textKnob('Button text', buttonTextPlaceholder);
	},

	toggleKnob(toggleDefault: boolean) {
		return booleanKnob('Toggle', toggleDefault);
	},

	disabledKnob(disabledDefault: boolean) {
		return booleanKnob('Disabled', disabledDefault);
	},

	formSectionHeaderKnob(headerPlaceholder: string) {
		return textKnob('Header text', headerPlaceholder);
	},

	formSectionSubmitKnob(submitPlaceholder: string) {
		return textKnob('Submit button text', submitPlaceholder);
	},

	tooltipAutoShowKnob(autoShowDefault: boolean) {
		return booleanKnob('Tooltip autoshow', autoShowDefault);
	},
};
