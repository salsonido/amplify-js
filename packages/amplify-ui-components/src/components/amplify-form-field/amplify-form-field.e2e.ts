import { newE2EPage, E2EPage } from '@stencil/core/testing';
import { pixelThreshold, selectors } from '../../common/testing';

describe('amplify-form-field e2e:', () => {
	let page: E2EPage;

	beforeEach(async () => {
		page = await newE2EPage();
	});

	it('renders with label, description, and hint', async () => {
		await page.setContent(
			`<amplify-form-field field-id="id" label="Label test" description="Description test" hint="Hint test" placeholder="Placeholder test"></amplify-form-field>`
		);
		const screenshot = await page.compareScreenshot('Amplify Form Field', {
			fullPage: true,
		});
		expect(screenshot).toMatchScreenshot({
			allowableMismatchedPixels: pixelThreshold,
		});

		const fieldElement = await page.find('amplify-form-field');
		expect(fieldElement).not.toBeNull();

		const labelElement = await page.find('label');
		expect(labelElement).toEqualText('Label test');
		expect(labelElement).toEqualAttribute('for', 'id');

		const descriptionElement = await page.find(selectors.formFieldDescription);
		expect(descriptionElement).toEqualText('Description test');
	});

	it('renders no label, description, or hint if none are provided', async () => {
		await page.setContent(`<amplify-form-field></amplify-form-field>`);

		const labelElement = await page.find('label');
		expect(labelElement).toBeNull();

		const descriptionElement = await page.find(selectors.formFieldDescription);
		expect(descriptionElement).toBeNull();
	});

	it('fires an onInputChange event when the contents of the box are changed', async () => {
		await page.setContent(`<amplify-form-field></amplify-form-field>`);

		const func = jest.fn();
		await page.exposeFunction('exposedfunc', func);

		await page.$eval('amplify-form-field', (fieldElement: any) => {
			fieldElement.handleInputChange = this.exposedfunc;
			fieldElement.label = 'adding a label so that the component rerenders';
		});
		await page.waitForChanges();

		const input = await page.find('input');
		await input.press('8');
		expect(func).toBeCalledTimes(1);
		expect(func.mock.calls[0][0].isTrusted).toBe(true);
		const value = await input.getProperty('value');
		expect(value).toBe('8');
	});

	it('can have a checkbox input', async () => {
		await page.setContent(
			`<amplify-form-field type='checkbox'></amplify-form-field>`
		);
		await page.waitForChanges();

		const input = await page.find('input');
		expect(input).toEqualAttribute('type', 'checkbox');
	});

	it('can have a number input', async () => {
		await page.setContent(
			`<amplify-form-field type='number'></amplify-form-field>`
		);
		await page.waitForChanges();

		const input = await page.find('input');
		expect(input).toEqualAttribute('type', 'number');
	});
});
