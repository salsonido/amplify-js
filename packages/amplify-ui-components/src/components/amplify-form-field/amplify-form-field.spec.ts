import { newSpecPage } from '@stencil/core/testing';
import { AmplifyFormField } from './amplify-form-field';

describe('amplify-form-field spec:', () => {
	describe('Component logic ->', () => {
		let formField;

		beforeEach(() => {
			formField = new AmplifyFormField();
		});

		it('should have fieldId prop be undefined by default', () => {
			expect(formField.fieldId).toBeUndefined();
		});

		it('should have label prop be undefined by default', () => {
			expect(formField.label).toBeUndefined();
		});

		it('should have description prop be undefined by default', () => {
			expect(formField.description).toBeUndefined();
		});

		it('should have type set to `text` by default', () => {
			expect(formField.type).toEqual('text');
		});

		it('should have required set to false by default', () => {
			expect(formField.required).toBe(false);
		});

		it('should have placeholder prop set to an empty string by default', () => {
			expect(formField.placeholder).toEqual('');
		});

		it('should have name prop be undefined by default', () => {
			expect(formField.name).toBeUndefined();
		});
	});
	describe('Render logic ->', () => {
		it('renders no label or description if none are provided', async () => {
			const page = await newSpecPage({
				components: [AmplifyFormField],
				html: `<amplify-form-field></amplify-form-field>`,
			});

			expect(page.root).toMatchSnapshot();
		});

		it('renders a label and description, if they are provided, and no id', async () => {
			const page = await newSpecPage({
				components: [AmplifyFormField],
				html: `<amplify-form-field label='label' description='description'></amplify-form-field>`,
			});

			expect(page.root).toMatchSnapshot();
		});

		it('renders with an id, if it is provided', async () => {
			const page = await newSpecPage({
				components: [AmplifyFormField],
				html: `<amplify-form-field label='label' description='description' field-id='id'></amplify-form-field>`,
			});

			expect(page.root).toMatchSnapshot();
		});

		it('renders with an name, if it is provided', async () => {
			const page = await newSpecPage({
				components: [AmplifyFormField],
				html: `<amplify-form-field label='label' description='description' field-id='id' name='seattle'></amplify-form-field>`,
			});

			expect(page.root).toMatchSnapshot();
		});

		it('replaces the input component, if a new one is provided', async () => {
			const page = await newSpecPage({
				components: [AmplifyFormField],
				html: `<amplify-form-field label='label' description='description' field-id='id'><input slot="input" /></amplify-form-field>`,
			});

			expect(page.root).toMatchSnapshot();
		});

		it('renders without Emotion CSS classes', async () => {
			const page = await newSpecPage({
				components: [AmplifyFormField],
				html: `<amplify-form-field label='label' description='description' field-id='id'></amplify-form-field>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});
});
