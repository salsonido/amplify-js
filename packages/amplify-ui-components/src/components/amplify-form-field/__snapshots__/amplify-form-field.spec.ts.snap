// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-form-field spec: Render logic -> renders a label and description, if they are provided, and no id 1`] = `
<amplify-form-field description="description" label="label">
  <!---->
  <div class="form-field">
    <div class="form-field-label">
      <amplify-label>
        label
      </amplify-label>
    </div>
    <div class="form-field-description" data-test="form-field-description" id="undefined-description">
      description
    </div>
    <div>
      <slot-fb name="input">
        <amplify-input description="description" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> renders no label or description if none are provided 1`] = `
<amplify-form-field>
  <!---->
  <div class="form-field">
    <div>
      <slot-fb name="input">
        <amplify-input placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> renders with an id, if it is provided 1`] = `
<amplify-form-field description="description" field-id="id" label="label">
  <!---->
  <div class="form-field">
    <div class="form-field-label">
      <amplify-label htmlfor="id">
        label
      </amplify-label>
    </div>
    <div class="form-field-description" data-test="form-field-description" id="id-description">
      description
    </div>
    <div>
      <slot-fb name="input">
        <amplify-input description="description" fieldid="id" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> renders with an name, if it is provided 1`] = `
<amplify-form-field description="description" field-id="id" label="label" name="seattle">
  <!---->
  <div class="form-field">
    <div class="form-field-label">
      <amplify-label htmlfor="id">
        label
      </amplify-label>
    </div>
    <div class="form-field-description" data-test="form-field-description" id="id-description">
      description
    </div>
    <div>
      <slot-fb name="input">
        <amplify-input description="description" fieldid="id" name="seattle" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> renders without Emotion CSS classes 1`] = `
<amplify-form-field description="description" field-id="id" label="label">
  <!---->
  <div class="form-field">
    <div class="form-field-label">
      <amplify-label htmlfor="id">
        label
      </amplify-label>
    </div>
    <div class="form-field-description" data-test="form-field-description" id="id-description">
      description
    </div>
    <div>
      <slot-fb name="input">
        <amplify-input description="description" fieldid="id" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> replaces the input component, if a new one is provided 1`] = `
<amplify-form-field description="description" field-id="id" label="label">
  <!---->
  <div class="form-field">
    <div class="form-field-label">
      <amplify-label htmlfor="id">
        label
      </amplify-label>
    </div>
    <div class="form-field-description" data-test="form-field-description" id="id-description">
      description
    </div>
    <div>
      <input slot="input">
      <slot-fb name="input">
        <amplify-input description="description" fieldid="id" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;
